runtime/parachains/src/paras_inherent.rs

```diff
@@ -868,169 +867,243 @@ fn sanitize_backed_candidates<
 	Ok(backed_candidates)
 }
 
-/// Limit the number of backed candidates processed in order to stay within block weight limits.
-///
-/// Use a configured assumption about the weight required to process a backed candidate and the
-/// current block weight as of the execution of this function to ensure that we don't overload
-/// the block with candidate processing.
-///
-/// If the backed candidates exceed the available block weight remaining, then skips all of them.
-/// This is somewhat less desirable than attempting to fit some of them, but is more fair in the
-/// even that we can't trust the provisioner to provide a fair / random ordering of candidates.
-fn limit_backed_candidates<T: Config>(
-	mut backed_candidates: Vec<BackedCandidate<T::Hash>>,
-) -> Vec<BackedCandidate<T::Hash>> {
-	const MAX_CODE_UPGRADES: usize = 1;
+fn limit_disputes<T: Config>(
+	disputes: &mut MultiDisputeStatementSet,
+) -> Weight {
+	let mut remaining_weight = <T as frame_system::Config>::BlockWeights::get().max_block;
+	let disputes_weight = dispute_statements_weight::<T>(&disputes);
+	if disputes_weight > remaining_weight {
+		// Sort the dispute statements according to the following prioritization:
+		//  1. Prioritize local disputes over remote disputes.
+		//  2. Prioritize older disputes over newer disputes.
+		disputes.sort_by(|a, b| {
+			let a_local_block = T::DisputesHandler::included_state(a.session, a.candidate_hash);
+			let b_local_block = T::DisputesHandler::included_state(b.session, b.candidate_hash);
+			match (a_local_block, b_local_block) {
+				// Prioritize local disputes over remote disputes.
+				(None, Some(_)) => Ordering::Greater,
+				(Some(_), None) => Ordering::Less,
+				// For local disputes, prioritize those that occur at an earlier height.
+				(Some(a_height), Some(b_height)) => a_height.cmp(&b_height),
+				// Prioritize earlier remote disputes using session as rough proxy.
+				(None, None) => a.session.cmp(&b.session),
+			}
+		});
 
-	// Ignore any candidates beyond one that contain code upgrades.
-	//
-	// This is an artificial limitation that does not appear in the guide as it is a practical
-	// concern around execution.
-	{
-		let mut code_upgrades = 0;
-		backed_candidates.retain(|c| {
-			if c.candidate.commitments.new_validation_code.is_some() {
-				if code_upgrades >= MAX_CODE_UPGRADES {
-					return false;
-				}
+		disputes.retain(|d| {
+			let dispute_weight = <<T as Config>::WeightInfo as WeightInfo>::enter_variable_disputes(d.statements.len() as u32);
+			let inclusion_weight = if remaining_weight > dispute_weight {
+				dispute_weight
+			} else {
+				0 as Weight
+			};
 
-				code_upgrades += 1;
+			if inclusion_weight > 0 {
+				remaining_weight -= dispute_weight;
 			}
 
-			true
+			inclusion_weight > 0
```

Comment:
This logic can be simplified to about:

```rust
let dispute_weight = <<T as Config>::WeightInfo as WeightInfo>::enter_variable_disputes(d.statements.len() as u32);

if remaining_weight > dispute_weight {
    remaining_weight -= dispute_weight
    true
} else {
    false
}
``

Reply:
We should prioritize local over remote ones, but also  randomize the remaining picks to not become gamable.

Reply:
Ok so basically we could

- iterate through disputes creating two new vecs: sorted_local_disputes & remote_disputes
- select local disputes until weight limit
- randomly select from `remote_disputes` until weight limit is reached

Reply:
Disputes are already sorted to prioritize local disputes in `fn limit_disputes`.

I can add the logic for selecting the remote disputes at random